### PR TITLE
Fix metabase-lib Metric class types

### PR DESCRIPTION
**Pull request by daltojohnso**
Related to https://github.com/metabase/metabase/issues/21689

This PR adds types to metabase-lib's `Metric` class. I'm approaching this similar to how I approached https://github.com/metabase/metabase/pull/21686 -- the constructor is typed to accept properties from an `IMetric` interface, and the Metric class itself has types on it and no longer extends `Base`. The functionality of `Base` now exists on `Metric` itself, so the functionality of the class should remain the same.

I don't think there's much, if any, typed metrics code (or if there is it is in a file with a `// @ts-nocheck` at the top), so there weren't any type errors caused by these changes. I went ahead and moved/fixed tests and added a `Metric` instance mock for easier mocking down the line.

**Testing**
1. Create a metric -- `/admin/datamodel/metrics`
2. Use the metric when summarizing in the simple query builder mode.

https://user-images.githubusercontent.com/13057258/163874683-4002b1ef-1a55-4e40-b841-2a506806a50d.mov

### CODE REVIEW

File: frontend/src/metabase-types/types/Metric.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { TableId } from "./Table";
 
-export type MetricId = number;
+export type MetricId = number | string;
```
Comment (daltojohnso) [inline frontend/src/metabase-types/types/Metric.ts:3]: According to https://github.com/metabase/metabase/pull/21708/files#diff-235568e206ffee752d7db7d812a123b233a0c3bb7d1e749bfcef3bc87286ae12L47 there is a special case where the id can in fact be a string.

File: frontend/src/metabase-lib/lib/metadata/Metric.ts [hunk 2]
```diff
@@ -1,35 +1,80 @@
...
+  StructuredQuery as StructuredQueryType,
+  MetricAgg,
+} from "metabase-types/types/Query";
+
+import Aggregation from "../queries/structured/Aggregation";
+import StructuredQuery from "../queries/StructuredQuery";
+import Table from "./Table";
+import Metadata from "./Metadata";
+
+export interface IMetric {
```
Comment (ranquild) [inline frontend/src/metabase-lib/lib/metadata/Metric.ts:11]: Should this be moved to `metabase-types/api`?

**Review by ranquild — APPROVED** [review]
A comment about API types

File: frontend/src/metabase-types/types/Metric.ts (hunk 1 above)
Comment (ranquild) [inline frontend/src/metabase-types/types/Metric.ts:3]: Could this be moved to `metabase-types/api` as well? There are similar ID types for databases and collections.

Comment (daltojohnso) [reply frontend/src/metabase-types/types/Metric.ts]: Good idea -- moved the type definition to `metabase-types/api`

File: frontend/src/metabase-types/api/metric.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { StructuredQuery } from "metabase-types/types/Query";
+
+export type MetricId = number | string;
+
+export interface IMetric {
```
Comment (ranquild) [inline frontend/src/metabase-types/api/metric.ts:5]: Other types don't use `I` prefix, it would be great to avoid that as well.

**Review by ranquild — APPROVED** [review]

**Pull request closed without merging**